src/backend/renderer/damage/shaper.rs

```diff
@@ -0,0 +1,407 @@
+                    });
+                }
+                in_damage[0].loc.x + in_damage[0].size.w
+            }
+            DamageSplitAxis::Y => {
+                if Some(direction) != last_direction {
+                    in_damage.sort_unstable_by(|lhs, rhs| {
+                        // Sort ascending by Y and then descending by height, so the first
+                        // rectangle will overlap the most.
+                        lhs.loc.y.cmp(&rhs.loc.y).then(rhs.size.h.cmp(&lhs.size.h))
+                    });
+                }
+                in_damage[0].loc.y + in_damage[0].size.h
+            }
+        };
+
+        // The start of overlap.
+        let mut overlap_start_idx = 0;
+        for idx in overlap_start_idx + 1..in_damage.len() {
+            let rect = in_damage[idx];
+            let (rect_start, rect_end) = match direction {
+                DamageSplitAxis::X => (rect.loc.x, rect.loc.x + rect.size.w),
+                DamageSplitAxis::Y => (rect.loc.y, rect.loc.y + rect.size.h),
+            };
+
+            // NOTE the renderer excludes the boundary, otherwise we need `>`.
+            if rect_start >= overlap_end {
+                self.shape_damage_impl(&mut in_damage[overlap_start_idx..idx], Some(direction), false);
+
+                // Advance the overlap.
+                overlap_start_idx = idx;
+                overlap_end = rect_end;
+            } else {
+                overlap_end = overlap_end.max(rect_end);
+            }
+        }
+
+        // When rectangle covers the entire bounding box and we've tried different direction of
+        // splitting perform the .
+        if overlap_start_idx == 0 && invert_direction {
+            // We pick more steps for edges which don't have full overlap.
+            const NUM_TILES: i32 = 4;
+            // NOTE we need to revert direction back to use larger side preferences.
+            let (tile_w, tile_h) = match direction.invert() {
+                DamageSplitAxis::X => (bbox_w / NUM_TILES, bbox_h / (NUM_TILES * 2)),
+                DamageSplitAxis::Y => (bbox_w / (NUM_TILES * 2), bbox_h / NUM_TILES),
+            };
+            let tile_size = (tile_w.max(MIN_TILE_SIDE), tile_h.max(MIN_TILE_SIDE)).into();
+
+            self.shape_damage_tiled(in_damage, damage_bbox, tile_size);
+        } else {
+            self.shape_damage_impl(
+                &mut in_damage[overlap_start_idx..],
+                Some(direction),
+                overlap_start_idx == 0,
+            );
+        }
+    }
+
+    #[inline]
+    fn shape_damage_tiled(
+        &mut self,
+        in_damage: &[Rectangle<i32, Physical>],
+        bbox: Rectangle<i32, Physical>,
+        tile_size: Size<i32, Physical>,
+    ) {
+        self.tiles_cache.clear();
+        let tile_gap: Size<i32, Physical> = From::from((
+            tile_size.h / TILE_DAMAGE_GAP_FRACTION,
+            tile_size.w / TILE_DAMAGE_GAP_FRACTION,
```

Comment:
Is it intended that h and w are swapped here?

Reply:
nope.